librz/analysis/arch/x86/x86_il.c

```diff
@@ -0,0 +1,3806 @@
+	X86Reg reg;
+	switch (size) {
+	case 8:
+		reg = X86_REG_AL;
+		break;
+	case 16:
+		reg = X86_REG_AX;
+		break;
+	case 32:
+		reg = X86_REG_EAX;
+		break;
+	case 64:
+		reg = X86_REG_RAX;
+		break;
+	default:
+		rz_warn_if_reached();
+		return NULL;
+	}
+
+	if (analysis->bits == 64) {
+		X86Reg mem_reg = X86_REG_RSI;
+		ut8 mem_size = 64;
+		/* Address override prefix: 67H */
+		if (ins->structure->prefix[3]) {
+			mem_reg = X86_REG_ESI;
+			mem_size = 32;
+		}
+
+		/* Cast to 64 if necessary (needed when address override prefix present) */
+		RzILOpPure *val = LOADW(size, (mem_size == 64 ? x86_il_get_reg(mem_reg) : UNSIGNED(64, x86_il_get_reg(mem_reg))));
+		RzILOpEffect *inc = x86_il_set_reg(mem_reg, ADD(x86_il_get_reg(mem_reg), UN(mem_size, size / BITS_PER_BYTE)));
+		RzILOpEffect *dec = x86_il_set_reg(mem_reg, SUB(x86_il_get_reg(mem_reg), UN(mem_size, size / BITS_PER_BYTE)));
+		RzILOpEffect *update_rsi = BRANCH(VARG(EFLAGS(DF)), dec, inc);
+
+		return SEQ2(x86_il_set_reg(reg, val), update_rsi);
+
+	} else {
+		X86Reg mem_reg = X86_REG_ESI;
+		ut8 mem_size = 32;
+		/* Address override prefix: 67H */
+		if (analysis->bits == 16 || ins->structure->prefix[3]) {
+			mem_reg = X86_REG_SI;
+			mem_size = 16;
+		}
+
+		X86Mem src_mem;
+		src_mem.base = mem_reg;
+		src_mem.disp = 0;
+		src_mem.index = X86_REG_INVALID;
+		src_mem.scale = 1;
+		src_mem.segment = X86_REG_DS;
+
+		/* No need for casting memaddr here since the casting will be done while calculating the segmented address */
+		RzILOpPure *val = LOADW(size, x86_il_get_memaddr(src_mem));
+
+		RzILOpEffect *inc = x86_il_set_reg(mem_reg, ADD(x86_il_get_reg(mem_reg), UN(mem_size, size / BITS_PER_BYTE)));
+		RzILOpEffect *dec = x86_il_set_reg(mem_reg, SUB(x86_il_get_reg(mem_reg), UN(mem_size, size / BITS_PER_BYTE)));
+		RzILOpEffect *update_si = BRANCH(VARG(EFLAGS(DF)), dec, inc);
+
+		return SEQ2(x86_il_set_reg(reg, val), update_si);
+	}
+}
+
+/**
+ * LODSB
+ * Load string byte
+ * No operands
+ */
+IL_LIFTER(lodsb) {
+	return x86_il_lods_helper(ins, pc, analysis, 8);
+}
+
+/**
+ * LODSW
+ * Load string word
+ * No operands
+ */
+IL_LIFTER(lodsw) {
+	return x86_il_lods_helper(ins, pc, analysis, 16);
+}
+
+/**
+ * LODSD
+ * Load string dword
+ * No operands
+ */
+IL_LIFTER(lodsd) {
+	return x86_il_lods_helper(ins, pc, analysis, 32);
+}
+
+/**
+ * LODSQ
+ * Load string quadword
+ * No operands
+ */
+IL_LIFTER(lodsq) {
+	return x86_il_lods_helper(ins, pc, analysis, 64);
+}
+
+// /**
+//  * LOOP
+//  * Loop the following instruction
+//  * Encoding: D
+//  * Decrement count ; jump if count != 0
+//  */
+// IL_LIFTER(loop) {
+// 	/* TODO: Unimplemented */
+// 	/* This is hard to implement currently, since it would require to loop the following instruction
+// 	and the IL currently provides no way to get the effect of another instruction */
+// 	/* Also, it's usage is usually very rare, so not worth all the effort. For now an `EMPTY` opcode is returned */
```

Comment:
It wasn't rare for DOS binaries

Reply:
Agreed. It seems fairly simple to implement now (I have no idea why I did not implement it in the first place). I'll add the implementation for `LOOP`, `LOOPE` and `LOOPNE` instructions.